gobblin-api/src/main/java/gobblin/testing/AssertWithBackoff.java

```diff
@@ -0,0 +1,200 @@
+   * needs for something asynchronous to happen, say a service to start.
+   *
+   * @param condition           the condition to wait for
+   * @param assertTimeoutMs     the max time in milliseconds to wait for the condition to become true
+   * @param message             the message to print while waiting for the condition
+   * @param log                 the logger to use for logging waiting, results
+   * @param backoffFactor       the number to multiple the sleep after condition failure;
+   *                            default is {@link #DEFAULT_ASSERT_BACKOFF_FACTOR}
+   * @param maxSleepMs          the max time to sleep between condition failures; default is
+   * @throws TimeoutException   if the condition did not become true in the specified time budget
+   * @throws InterrupedException if the assert gets interrupted while waiting for the condition to
+   *                            become true.
+   */
+  public static void assertTrue(
+      Predicate<Void> condition, long assertTimeoutMs, String message,
+      Logger log, double backoffFactor, long maxSleepMs)
+          throws TimeoutException, InterruptedException {
+    long startTimeMs = System.currentTimeMillis();
+    long endTimeMs = startTimeMs + assertTimeoutMs;
+    long currentSleepMs = 0;
+    boolean done = false;
+    try {
+      while (!done && System.currentTimeMillis() < endTimeMs) {
+          done = condition.apply(null);
+          if (!done) {
+            currentSleepMs = computeRetrySleep(currentSleepMs, backoffFactor, maxSleepMs,
+                endTimeMs);
+            log.debug("Condition check for '" + message + "' failed; sleeping for " +
+                      currentSleepMs + "ms");
+            Thread.sleep(currentSleepMs);
+          }
+      }
+      //one last try
+      if (!done && !condition.apply(null)) {
+        throw new TimeoutException("Timeout waiting for condition '" + message + "'.");
+      }
+    } catch (TimeoutException|InterruptedException e) {
+      //pass through
+      throw e;
+    } catch (RuntimeException e) {
+      throw new RuntimeException("Exception checking condition '" + message + "':" + e, e);
+    }
+  }
+
+  /**
+   * Computes a new sleep for a retry of a condition check.
+   *
+   * @param currentSleepMs    the last sleep duration in milliseconds
+   * @param backoffFactor     the factor to multiple currentSleepMs
+   * @param maxSleepMs        the maximum allowed sleep in milliseconds
+   * @param endTimeMs         the end time based on timeout for the condition to become true
+   * @return the new sleep time which will not exceed maxSleepMs and will also not cause to
+   *         overshoot endTimeMs
+   */
+  public static long computeRetrySleep(long currentSleepMs, double backoffFactor, long maxSleepMs,
+                                       long endTimeMs) {
+    long newSleepMs = Math.round(currentSleepMs * backoffFactor);
+    if (newSleepMs <= currentSleepMs) {
+      //prevent infinite loops
```

Comment:
First letter capital and one empty space after `//`, to be consistent with the rest of the code base.